File: OrangeJuice.Server.Core/Configuration/CachingAwsOptionsProvider.cs
using System.Collections.Generic;
using System.Threading.Tasks;

using OrangeJuice.Server.Cache;

namespace OrangeJuice.Server.Configuration
{
	public class CachingAwsOptionsProvider : IOptionsProvider<AwsOptions>
	{
		private readonly IOptionsProvider<AwsOptions> _optionsProvider;
		private readonly ICacheClient _cacheClient;

		public CachingAwsOptionsProvider(IOptionsProvider<AwsOptions> optionsProvider, ICacheClient cacheClient)
		{
			_optionsProvider = optionsProvider;
			_cacheClient = cacheClient;
		}

		public Task<IEnumerable<AwsOptions>> GetOptions()
		{
			return _cacheClient.AddOrGetExisting(AwsOptions.CacheKey, () => _optionsProvider.GetOptions());
		}
	}
}

File: OrangeJuice.Server.Core.FSharp/Configuration/CachingAwsOptionsProvider.fs
namespace OrangeJuice.Server.FSharp.Configuration

open System.Threading.Tasks

open OrangeJuice.Server.Cache
open OrangeJuice.Server.Configuration

type CachingAwsOptionsProvider(optionsProvider : IOptionsProvider<AwsOptions>, cacheClient : ICacheClient) =
    interface IOptionsProvider<AwsOptions> with
        member this.GetOptions() : Task<AwsOptions[]> =
            cacheClient.AddOrGetExisting(AwsOptions.CacheKey, fun() -> optionsProvider.GetOptions())
